routing/payment_lifecycle.go

```diff
@@ -1095,3 +1062,100 @@ func (p *paymentLifecycle) refreshPayment() (DBMPPayment,
 
 	return payment, ps, nil
 }
+
+// processSwitchResults reads the `p.results` map and process the results
+// returned from the htlcswitch.
+func (p *paymentLifecycle) processSwitchResults() error {
+	// Create a slice to remember the results of the attempts that we have
+	// processed.
+	attempts := make([]*channeldb.HTLCAttempt, 0, p.switchResults.Len())
+
+	var errReturned error
+
+	// Range over the map to process all the results.
+	p.switchResults.Range(func(a *channeldb.HTLCAttempt,
+		result *htlcswitch.PaymentResult) bool {
+
+		// Save the keys so we know which items to delete from the map.
+		attempts = append(attempts, a)
+
+		// Handle the attempt result. If an error is returned here, it
+		// means the payment lifecycle needs to be terminated.
+		_, err := p.handleAttemptResult(a, result)
+		if err != nil {
+			errReturned = err
+		}
+
+		// Always return true so we will process all results.
+		return true
+	})
+
+	// Clean up the processed results.
+	for _, a := range attempts {
+		p.switchResults.Delete(a)
+	}
+
+	return errReturned
+}
+
+// handleAttemptResult processes the result of an HTLC attempt returned from
+// the htlcswitch.
+func (p *paymentLifecycle) handleAttemptResult(attempt *channeldb.HTLCAttempt,
+	result *htlcswitch.PaymentResult) (*attemptResult, error) {
+
+	// In case of a payment failure, fail the attempt with the control
+	// tower and return.
+	if result.Error != nil {
+		return p.handleSwitchErr(attempt, result.Error)
```

Comment:
I wonder if the payment is failed here, we do also at least report this via the `trackpayment` stream which might be registered for this payment ?